### PR TITLE
secp256k1: Add optimized sqrt field calc.

### PR DESCRIPTION
**Pull request by davecgh**
This introduces the ability to calculate the square root modulo the field characteristic, when it exists, via the specialized field value type.  It is significantly faster than performing the calculation via big integers and it is also constant time.

The motivation for adding this is that it is needed for point decompression which currently makes use of big ints.

Comprehensive tests with 100% coverage and a full suite of benchmarks are included.

Like other code in this area, the code relies on techniques that are not the easiest to review for correctness, so significant effort has been put into attempting to clearly explain the details via comments to help with review.

For now, this merely introduces the method without modifying any of the code to make use of it.  The intention is to do so in future commits.

The following benchmark shows a comparison between calculating the square root via the new specialized method and via generic big ints.  As can be seen, the new method is roughly twice as fast and  reduces allocations by 100%.

```
benchmark       old ns/op    new ns/op    delta
--------------------------------------------------
BenchmarkSqrt   24605        11087        -54.94%

benchmark       old allocs   new allocs   delta
--------------------------------------------------
BenchmarkSqrt   91           0            -100.00%

benchmark       old bytes    new bytes    delta
--------------------------------------------------
BenchmarkSqrt   1016         0            -100.00%
```

### CODE REVIEW

File: dcrec/secp256k1/field.go [hunk 1]
```diff
@@ -894,6 +894,165 @@ func (f *fieldVal) Mul2(val *fieldVal, val2 *fieldVal) *fieldVal {
...
+	f.Square().Square().Square().Square().Square()         // f = a^(2^238 - 2^15)
+	f.Square().Square().Square().Square().Square()         // f = a^(2^243 - 2^20)
+	f.Square().Square().Square()                           // f = a^(2^246 - 2^23)
+	f.Mul(&a22)                                            // f = a^(2^246 - 2^22 - 1)
+	f.Square().Square().Square().Square().Square()         // f = a^(2^251 - 2^27 - 2^5)
+	f.Square()                                             // f = a^(2^252 - 2^28 - 2^6)
+	f.Mul(&a2)                                             // f = a^(2^252 - 2^28 - 2^6 - 2^1 - 1)
+	f.Square().Square()                                    // f = a^(2^254 - 2^30 - 2^8 - 2^3 - 2^2)
+	//                                                     //   = a^(2^254 - 2^30 - 244)
+	//                                                     //   = a^((p+1)/4)
```
Comment (matheusd) [inline dcrec/secp256k1/field.go]: Ok, I spent waaaaay longer on this than I should have and I haven't internalized where this specific set operations came from. But I've verified that all the provided equations hold via wolfram alpha and it computes what it claims to.

`[a3] a^(2^3-1) == (a^(2^2-1))^2 * a`
https://www.wolframalpha.com/input/?i=a%5E%282%5E3-1%29+%3D%3D+%28a%5E%282%5E2-1%29%29%5E2+*+a

`[a6] a^(2^6-1) = (a^(2^3-1))^(2^3) * a^(2^3-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E6-1%29+%3D+%28a%5E%282%5E3-1%29%29%5E%282%5E3%29+*+a%5E%282%5E3-1%29

`[a9] a^(2^9-1) = (a^(2^6-1))^(2^3) * a^(2^3-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E9-1%29+%3D+%28a%5E%282%5E6-1%29%29%5E%282%5E3%29+*+a%5E%282%5E3-1%29

`[a11] a^(2^11-1) = (a^(2^9-1))^(2^2) * a^(2^2-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E11-1%29+%3D+%28a%5E%282%5E9-1%29%29%5E%282%5E2%29+*+a%5E%282%5E2-1%29

`[a22] a^(2^22-1) = (a^(2^11-1))^(2^11) * a^(2^11-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E22-1%29+%3D+%28a%5E%282%5E11-1%29%29%5E%282%5E11%29+*+a%5E%282%5E11-1%29

`[a44] a^(2^44-1) = (a^(2^22-1))^(2^22) * a^(2^22-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E44-1%29+%3D+%28a%5E%282%5E22-1%29%29%5E%282%5E22%29+*+a%5E%282%5E22-1%29

`[a88] a^(2^88-1) = (a^(2^44-1))^(2^44) * a^(2^44-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E88-1%29+%3D+%28a%5E%282%5E44-1%29%29%5E%282%5E44%29+*+a%5E%282%5E44-1%29

`[a176] a^(2^176-1) = (a^(2^88-1))^(2^88) * a^(2^88-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E176-1%29+%3D+%28a%5E%282%5E88-1%29%29%5E%282%5E88%29+*+a%5E%282%5E88-1%29

`[a220] a^(2^220-1) = (a^(2^176-1))^(2^44) * a^(2^44-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E220-1%29+%3D+%28a%5E%282%5E176-1%29%29%5E%282%5E44%29+*+a%5E%282%5E44-1%29

`[a223] a^(2^223-1) = (a^(2^220-1))^(2^3) * a^(2^3-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E223-1%29+%3D+%28a%5E%282%5E220-1%29%29%5E%282%5E3%29+*+a%5E%282%5E3-1%29

`[a(p+1/4)] a^(2^254-2^30-244) = ((((a^(2^223-1))^(2^23) * a^(2^22-1))^(2^6)) * a^(2^2-1))^2^2`
https://www.wolframalpha.com/input/?i=a%5E%282%5E254-2%5E30-244%29+%3D+%28%28%28%28a%5E%282%5E223-1%29%29%5E%282%5E23%29+*+a%5E%282%5E22-1%29%29%5E%282%5E6%29%29+*+a%5E%282%5E2-1%29%29%5E2%5E2

File: dcrec/secp256k1/field.go [hunk 2]
```diff
@@ -894,6 +894,165 @@ func (f *fieldVal) Mul2(val *fieldVal, val2 *fieldVal) *fieldVal {
...
+	// 00111111 11111111 11111111 11111111
+	// 11111111 11111111 11111111 11111111
+	// 11111111 11111111 11111111 11111111
+	// 11111111 11111111 11111111 11111111
+	// 11111111 11111111 11111111 11111111
+	// 11111111 11111111 11111111 11111111
+	// 11111111 11111111 11111111 11111111
+	// 10111111 11111111 11111111 00001100
+	//
+	// Notice that can be broken up into five windows of consecutive 1s (in
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/field.go]: Indeed, it is three as specifically detailed just below.  Fixed.

File: dcrec/secp256k1/field.go [hunk 3]
```diff
@@ -894,6 +894,165 @@ func (f *fieldVal) Mul2(val *fieldVal, val2 *fieldVal) *fieldVal {
...
+	// Further, Euler's criterion states that an integer 'a' has a square root
+	// (aka is a quadratic residue) modulo a prime if a^((p-1)/2) ≡ 1 (mod p)
+	// and, conversely, when it does NOT have a square root (aka is 'a'
+	// non-residue) a^((p-1)/2) ≡ -1 (mod p).
+	//
+	// This can be seen by considering that Fermat's little theorem can be
+	// written as (a^((p-1)/2) - 1)(a^((p-1)/2) + 1) ≡ 0 (mod p).  Therefore,
+	// one of the two factors must be 0.  Then, when a ≡ x^2 (aka 'a' is a
+	// quadratic residue), (x^2)^((p-1)/2) ≡ x^(p-1) ≡ 1 (mod p) which implies
+	// the first factor must be zero.  Finally, per Legrange's theorem, the
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/field.go]: Fixed.

File: dcrec/secp256k1/field.go (hunk 1 above)
Comment (davecgh) [reply dcrec/secp256k1/field.go]: The intuition is that it's just summing the partial products.  Or stated differently, it's breaking up the multiplication into sum of easier multiplications.

For example, imagine you have the number 123045067 in decimal.  It's essentially just doing 123*1000000 + 45*1000 + 67*1 = 123045067, except in base 2 and with much much larger numbers.

**Pull request merged**